keps/sig-auth/4317-pod-certificates/README.md

```diff
@@ -0,0 +1,1294 @@
+	// that this list may be expanded in the future.
+	//
+	// Signer implementations do not need to support all key types supported by
+	// kube-apiserver and kubelet.  If a signer does not support the key type
+	// used for a given PodCertificateRequest, it should deny the request, with
+	// a reason of UnsupportedKeyType.  It may also suggest a key type that it
+	// does support by attaching an additional SuggestedKeyType condition, with
+	// its reason field set to the suggested key type identifier.
+	PKIXPublicKey []byte `json:"pkixPublicKey" protobuf:"bytes,8,opt,name=pkixPublicKey"`
+
+	// proofOfPossession proves that the requesting Kubelet holds the private
+	// key corresponding to pkixPublicKey.
+	//
+	// kube-apiserver validates the proof of possession during creation of the
+	// PodCertificateRequest.
+	//
+	// If the key is an RSA key, then the signature is over the ASCII bytes of
+	// the pod UID, using RSASSA-PKCS1-V1_5-SIGN from RSA PKCS #1 v1.5 (as
+	// implemented by the golang function crypto/rsa.SignPKCS1v15).
+	//
+	// If the key is an ECDSA key, then the signature is as described by [SEC 1,
+	// Version 2.0](https://www.secg.org/sec1-v2.pdf) (as implemented by the
+	// golang library function crypto/ecdsa.SignASN1)
+	ProofOfPossession []byte `json:"proofOfPossession" protobuf:"bytes,9,opt,name=proofOfPossession"`
+}
+
+type PodCertificateRequestStatus struct {
+	// conditions applied to the request. Known conditions are "Denied",
+	// "Failed", and "SuggestedKeyType".
+	//
+	// +listType=map
+	// +listMapKey=type
+	// +optional
+	Conditions []metav1.Condition `json:"conditions,omitempty" protobuf:"bytes,1,rep,name=conditions"`
+
+	// certificateChain is populated with an issued certificate by the signer.
+	// This field is set via the /status subresource. Once populated, this field
+	// is immutable.
+	//
+	// If the certificate signing request is denied, a condition of type
+	// "Denied" is added and this field remains empty. If the signer cannot
+	// issue the certificate, a condition of type "Failed" is added and this
+	// field remains empty.
+	//
+	// Validation requirements:
+	//  1. certificateChain must contain one or more entries.
+	//  2. Each entry must contain a single DER-encoded ASN.1 Certificate structure as described in section 4 of RFC5280.
+	//
+	// If more than one entry is present, and the definition of the requested
+	// spec.signerName does not indicate otherwise, the first entry is the
+	// issued certificate, and subsequent entries should be treated as
+	// intermediate certificates and presented in TLS handshakes.  When
+	// projecting the chain into a pod volume, kubelet will preserve the order
+	// of certificateChain.
+	//
+	// As a reminder, when serialized as JSON or YAML, each entry is
+	// additionally base64-encoded.
+	//
+	// +listType=atomic
+	// +optional
+	CertificateChain [][]byte `json:"certificateChain,omitempty" protobuf:"bytes,2,opt,name=certificateChain"`
```

Comment:
If the kubelet will need to be able to project intermediates distinct from the leaf cert, having `certificate []byte` and `intermediates [][]byte` might make more sense

If the kubelet will only inject PEM format, is there a benefit to making this different than the ultimately injected value? Everywhere else we have certificates in the REST API or config types we use PEM format (either directly as string or base64-encoded as []byte)

Reply:
There's no use case for projecting the cert separate from intermediates.  A properly-functioning client will always present them all, in the same order that they were projected into the filesystem.

As for PEM vs raw-DER, my thinking was that humans are never going to be requesting these certificates, so there's not much point in using PEM for the intermediate representation.  So we might as well save on some unneeded base64 round-tripping.

I don't fell strongly either way... this field could also be `CertificateChain string`, and just contain a bunch of PEM blocks.

Reply:
Done.